packages/next-swc/crates/core/src/relay.rs

```diff
@@ -0,0 +1,284 @@
+    })
+}
+
+impl Fold for Relay {
+    fn fold_expr(&mut self, expr: Expr) -> Expr {
+        let expr = expr.fold_children_with(self);
+
+        match &expr {
+            Expr::TaggedTpl(tpl) => {
+                if let Some(built_expr) = self.build_call_expr_from_tpl(tpl) {
+                    built_expr
+                } else {
+                    expr
+                }
+            }
+            _ => expr,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum BuildRequirePathError<'a> {
+    FileNameNotReal,
+    MultipleSourceSetsFound {
+        source_set_names: Vec<SourceSetName>,
+        path: &'a PathBuf,
+    },
+    ProjectNotFoundForSourceSet {
+        source_set_name: SourceSetName,
+    },
+    FileNotASourceFile,
+    CouldNotCategorize {
+        err: Cow<'static, str>,
+        path: String,
+    },
+}
+
+// This is copied from https://github.com/facebook/relay/blob/main/compiler/crates/relay-compiler/src/build_project/generate_artifacts.rs#L251
+// until the Relay team exposes it for external use.
+fn path_for_artifact(
+    root_dir: &PathBuf,
+    source_path: &PathBuf,
+    project_config: &ProjectConfig,
+    definition_name: String,
+) -> PathBuf {
+    let source_file_location_key = SourceLocationKey::Standalone {
+        path: source_path.to_str().unwrap().parse().unwrap(),
+    };
+    let filename = if let Some(filename_for_artifact) = &project_config.filename_for_artifact {
+        filename_for_artifact(source_file_location_key, definition_name.parse().unwrap())
+    } else {
+        match &project_config.typegen_config.language {
+            relay_config::TypegenLanguage::Flow => format!("{}.graphql.js", definition_name),
+            relay_config::TypegenLanguage::TypeScript => {
+                format!("{}.graphql.ts", definition_name)
+            }
+        }
+    };
+
+    let output_path = create_path_for_artifact(project_config, source_file_location_key, filename);
+
+    if project_config.output.is_some() {
+        let absolute_output_path = root_dir.join(&output_path);
+
+        let diffed_path =
+            diff_paths(&absolute_output_path, &source_path.parent().unwrap()).unwrap();
+
+        return diffed_path;
+    }
+
+    output_path
+}
+impl Relay {
+    fn build_require_path(
+        &mut self,
+        operation_name: String,
+    ) -> Result<PathBuf, BuildRequirePathError> {
+        match &self.relay_config_for_tests {
+            Some(config) => match &self.file_name {
+                FileName::Real(real_file_path) => Ok(path_for_artifact(
+                    &std::env::current_dir().unwrap(),
+                    &real_file_path,
+                    config,
+                    operation_name,
+                )),
+                _ => Err(BuildRequirePathError::FileNameNotReal),
+            },
+            _ => {
+                let config =
+                    relay_compiler::config::Config::search(&std::env::current_dir().unwrap())
+                        .unwrap();
```

Comment:
I think it is a reasonable to support only this type of loading config, but should we add a better error message (for the second `unwrap`) in case the config was not found?

Reply:
@alunyov What other config loading style is there? I don't mind adding support for something else.

Reply:
Now, when I'm thinking about it, I don't think we can support other ways of lading configs here.

Maybe, in this context, we will be able to handle only two ways:
- `Config::search` (already added) and `Config::load` - when people explicitly passing the path to a JSON config file via CLI
- And also we have a limited set of CLI arguments that we can convert to a config:

For all these CLI options, not sure how we can translate them here.

For the reference:
https://github.com/facebook/relay/blob/d9d8783ec7c162dab1fea0f9049c892316192f74/compiler/crates/relay-compiler/src/main.rs#L97-L103

Reply:
Gotcha, thanks for the context. I think we can leave it as is, I would expect anyone using this to have their relay config at some ancestor of the cwd.

Reply:
Also, maybe we can address this in a follow up PR. I'm helping out a team with a Relay / GraphQL migration this week and want to keep them on the Rust compiler